Feature/login

k0vey opened the pull request:
You can signup and login as a user (create a user and you can test if he is logged on) (No roles yet)

k0vey committed:
Add authorization service

wizacass committed:
Update configuration and signup route to properly generate JWT token

k0vey committed:
Added users API

We can create users and log in as well as check if we logged in.

k0vey committed:
Fixed jwt autherization with roles

You can create users with roles and use and lock method to certain role users

wizacass commented on Controllers/UserController.cs:
```diff
@@ -0,0 +1,140 @@
+            _roleManager = roleManager;
+            _jwt = new JwtService(config);
+        }
+
+
```
Extra line here should be removed. There should be 1 line between methods.

wizacass commented on Controllers/UserController.cs:
```diff
@@ -0,0 +1,140 @@
+        [HttpPost("login")]
+        [AllowAnonymous]
+        public async Task<IActionResult> Login(AuthDTO model)
+        {
+            var user = await _userManager.FindByEmailAsync(model.Email);
```
Request validation check is missing. We always do this check first to check headers:
```C#
if (!IsValidApiRequest())
{
  return ApiBadRequest("Invalid Headers!");
}
```

wizacass commented on Controllers/UserController.cs:
```diff
@@ -0,0 +1,140 @@
+            if (!result.Succeeded)
+                return ApiBadRequest("Invalid username or password.");
+
+            List<string> roles = (List<string>) await _userManager.GetRolesAsync(user);
```
Can't we use `var here`?

wizacass commented on Controllers/UserController.cs:
```diff
@@ -0,0 +1,140 @@
+                return ApiBadRequest("Invalid username or password.");
+
+            List<string> roles = (List<string>) await _userManager.GetRolesAsync(user);
+
+
```
Extra newline.

wizacass commented on Controllers/UserController.cs:
```diff
@@ -0,0 +1,140 @@
+            var result = await _userManager.CreateAsync(user, model.Password);
+            if (!result.Succeeded)
+                return ApiBadRequest(result.Errors.First().Description);
+
+           
+
+            
```
Unecessary whitespace.

wizacass commented on Controllers/UserController.cs:
```diff
@@ -0,0 +1,140 @@
+        {
+            return Ok( HttpContext.User.Identity.Name);
+        }
+
```
Unecessary whitespace.

wizacass commented on Controllers/UserController.cs:
```diff
@@ -0,0 +1,140 @@
+                })
+            });
+        }
+
```
Unecessary whitespace.

wizacass commented on Controllers/UserController.cs:
```diff
@@ -0,0 +1,140 @@
+        [AllowAnonymous]
+        public async Task<IActionResult> Signup(AuthDTO model)
+        {
+            if(!(await _roleManager.RoleExistsAsync("0")))
+            {
+                await _roleManager.CreateAsync(new IdentityRole("0"));
+            }
+            if (!(await _roleManager.RoleExistsAsync("1")))
+            {
+                await _roleManager.CreateAsync(new IdentityRole("1"));
+            }
+            if (!(await _roleManager.RoleExistsAsync("2")))
+            {
+                await _roleManager.CreateAsync(new IdentityRole("2"));
+            }
+            if (!(await _roleManager.RoleExistsAsync("3")))
+            {
+                await _roleManager.CreateAsync(new IdentityRole("3"));
+            }
```
Can't we use `switch` here? Seems like this logic could be greatly simplified.

wizacass commented on Controllers/UserController.cs:
```diff
@@ -0,0 +1,140 @@
+
+            
+            var userFromDb = await _userManager.FindByNameAsync(user.UserName);
+            if (model.RoleId.ToString() == "0")
+                await _userManager.AddToRoleAsync(userFromDb, "0");
+            if (model.RoleId.ToString() == "1")
+                await _userManager.AddToRoleAsync(userFromDb, "1");
+            if (model.RoleId.ToString() == "2")
+                await _userManager.AddToRoleAsync(userFromDb, "2");
+            if (model.RoleId.ToString() == "3")
+                await _userManager.AddToRoleAsync(userFromDb, "3");
```
This logic should be simplified. It breaks DRY principle.

wizacass commented on Controllers/UserController.cs:
```diff
@@ -0,0 +1,140 @@
+
+        [Authorize(Roles = "0")]
+        [HttpGet("Me")]
+         public async Task<IActionResult> getMe()
```
Indentation error.

wizacass commented on Models/ApiContext.cs:
```diff
@@ -34,7 +39,28 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
             modelBuilder.ApplyConfiguration(new ProviderWarehouseConfiguration());
 
-            base.OnModelCreating(modelBuilder);
+            modelBuilder.Entity<IdentityUser>(entity => entity.Property(m => m.Id).HasMaxLength(85));
+            modelBuilder.Entity<IdentityUser>(entity => entity.Property(m => m.NormalizedEmail).HasMaxLength(85));
+            modelBuilder.Entity<IdentityUser>(entity => entity.Property(m => m.NormalizedUserName).HasMaxLength(85));
+
+            modelBuilder.Entity<IdentityRole>(entity => entity.Property(m => m.Id).HasMaxLength(85));
+            modelBuilder.Entity<IdentityRole>(entity => entity.Property(m => m.NormalizedName).HasMaxLength(85));
+
+            modelBuilder.Entity<IdentityUserLogin<string>>(entity => entity.Property(m => m.LoginProvider).HasMaxLength(85));
+            modelBuilder.Entity<IdentityUserLogin<string>>(entity => entity.Property(m => m.ProviderKey).HasMaxLength(85));
+            modelBuilder.Entity<IdentityUserLogin<string>>(entity => entity.Property(m => m.UserId).HasMaxLength(85));
+            modelBuilder.Entity<IdentityUserRole<string>>(entity => entity.Property(m => m.UserId).HasMaxLength(85));
+
+            modelBuilder.Entity<IdentityUserRole<string>>(entity => entity.Property(m => m.RoleId).HasMaxLength(85));
+
+            modelBuilder.Entity<IdentityUserToken<string>>(entity => entity.Property(m => m.UserId).HasMaxLength(85));
+            modelBuilder.Entity<IdentityUserToken<string>>(entity => entity.Property(m => m.LoginProvider).HasMaxLength(85));
+            modelBuilder.Entity<IdentityUserToken<string>>(entity => entity.Property(m => m.Name).HasMaxLength(85));
+
+            modelBuilder.Entity<IdentityUserClaim<string>>(entity => entity.Property(m => m.Id).HasMaxLength(85));
+            modelBuilder.Entity<IdentityUserClaim<string>>(entity => entity.Property(m => m.UserId).HasMaxLength(85));
+            modelBuilder.Entity<IdentityRoleClaim<string>>(entity => entity.Property(m => m.Id).HasMaxLength(85));
+            modelBuilder.Entity<IdentityRoleClaim<string>>(entity => entity.Property(m => m.RoleId).HasMaxLength(85));
```
It would be better to extract these rules to their own model configuration classes as it is done with other models.

wizacass commented on Models/DTO/AuthDTO.cs:
```diff
@@ -0,0 +1,23 @@
+        public string Password { get; set; }
+
+        [Required]
+        public int RoleId { get; set; }
```
Can't we use `DepartmentId` enum here instead?

wizacass commented on Models/JwtUser.cs:
```diff
@@ -0,0 +1,9 @@
+    {
+        public string Username { get; set; }
+        public string Email { get; set; }
+        public int roleId { get; set; }
```
Same here as in `AuthDTO`.

wizacass commented on Models/User.cs:
```diff
@@ -0,0 +1,7 @@
+
+namespace API.Models
+{
+    public class User : IdentityUser {
+    }
```
Maybe we should format this as `public class User : IdentityUser { }` because it is highly likely this class will stay like this for the duration of the project.

k0vey committed:
Cleaned up the code and created new configuration files

k0vey committed:
Corrected the user class

wizacass commented on Controllers/UserController.cs:
```diff
@@ -131,7 +135,7 @@ public async Task<IActionResult> Signup(AuthDTO model)
         [Authorize(Roles = "0")]
         [HttpGet("Me")]
-         public async Task<IActionResult> getMe()
+        public async Task<IActionResult> getMe()
```
Method name should start with a capital letter

k0vey committed:
Updated roleId variable from int to DepartmentId

Updated roleId variable from int to DepartmentId and tested the new model

Note: In signup you need to declare role as a number(1 to 4)
In methods [Authorize(Roles = "RoleName")] RoleName example Phamacy,Warehouse,Transportation,Admin

k0vey committed:
Fixed bugs in registration and logins

Made the code easier to read as well fixed some issues with logins and registration. Added swagger for testing purposes
